File: canFinish.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& points) {
        int visited = 0;
        vector<int> point(numCourses);
        for(int i=0;i<points.size();i++){
            point[points[i][1]]++;
        }
        auto iter = find(point.begin(), point.end(), 0);
        while(iter!=point.end()){
            visited++;
            for(int j=0;j<points.size();j++){
                if(points[j][0]==&*iter-&point[0]){ //查找元素下标的方式。
                    point[points[j][1]]--;
                }
            }
            *iter = -1;
            numCourses--;
            iter = find(point.begin(),point.end(),0);
        }
        if(numCourses<visited){return true;}
        else {return false;}
    }
};

File: canFinish.py
class Solution:
    # 拓扑排序的应用
    def canFinish(self, numCourses: int, points: List[List[int]]) -> bool:
        degree = [0 for i in range(numCourses)]
        visited = 0
        for point in points:
            degree[point[1]]+=1
        while 0 in degree:
            for i in range(len(degree)):
                if degree[i] == 0:
                    degree[i] = -1
                    visited += 1
                    for j in points:
                        if j[0] == i:
                            degree[j[1]] -= 1
                    numCourses -= 1
                else:
                    continue
        return True if numCourses<visited else False
